grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy

```diff
@@ -272,34 +272,51 @@ class ProductSupplierService {
 
         PreferenceType preferenceType = params.globalPreferenceTypeName ? PreferenceType.findByName(params.globalPreferenceTypeName) : null
 
+        assignDefaultPreferenceType(productSupplier,
+                preferenceType,
+                params.globalPreferenceTypeComments,
+                params.globalPreferenceTypeValidityStartDate,
+                params.globalPreferenceTypeValidityEndDate)
+
+        if (!productSupplier.code && !productSupplier.id) {
+            assignSourceCode(productSupplier, supplier)
+        }
+        return productSupplier
+    }
+
+    void assignDefaultPreferenceType(ProductSupplier productSupplier,
+                 PreferenceType preferenceType,
+                 String comments,
+                 String validityStartDate,
+                 String validityEndDate) {
+        ProductSupplierPreference productSupplierPreference = productSupplier.getGlobalProductSupplierPreference()
         if (preferenceType) {
-            ProductSupplierPreference productSupplierPreference = productSupplier.getGlobalProductSupplierPreference()
+            SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy")
 
             if (!productSupplierPreference) {
                 productSupplierPreference = new ProductSupplierPreference()
                 productSupplier.addToProductSupplierPreferences(productSupplierPreference)
             }
 
             productSupplierPreference.preferenceType = preferenceType
-            productSupplierPreference.comments = params.globalPreferenceTypeComments
+            productSupplierPreference.comments = comments
 
-            def globalPreferenceTypeValidityStartDate = params.globalPreferenceTypeValidityStartDate ? dateFormat.parse(params.globalPreferenceTypeValidityStartDate) : null
+            Date globalPreferenceTypeValidityStartDate = validityStartDate ? dateFormat.parse(validityStartDate) : null
 
             if (globalPreferenceTypeValidityStartDate) {
                 productSupplierPreference.validityStartDate = globalPreferenceTypeValidityStartDate
             }
 
-            def globalPreferenceTypeValidityEndDate = params.globalPreferenceTypeValidityEndDate ? dateFormat.parse(params.globalPreferenceTypeValidityEndDate) : null
+            Date globalPreferenceTypeValidityEndDate = validityEndDate ? dateFormat.parse(validityEndDate) : null
 
             if (globalPreferenceTypeValidityEndDate) {
                 productSupplierPreference.validityEndDate = globalPreferenceTypeValidityEndDate
             }
+            return
```

Comment:
could you elaborate what would be cleaner? in my opinion doing `if (!preferenceType)` at the beginning wouldn't change much, because the logic of assigning preferences would have to stay the same, so the only change would be, that this logic wouldn't be nested inside the `if` statement, unless I miss something :thinking: